openshift/assisted-image-service
# MGMT-9333 Add ipxe-initrd endpoint for customized initrd

## Description


This PR creates a new endpoint in the image service which will serve an initrd which includes the ignition for a particular infra-env.

The new endpoint is of the form:

```
/images/{image-id}/pxe-initrd?{params}
```

For example:

```
/images/dbe345ff-a89e-4c61-878f-d50066d1af3a/pxe-initrd?version=4.9&arch=x86_64
```

The restructuring of the image handler is worth calling out specifically.
The go standard library http server implementation works only on path prefixes.
This means that to implement a sub-path of a particular resource (such as this one) we need an intermediate handler that handles parsing the path and calling the correct sub-handler.

This PR turns the image handler into this "base" handler that will route to either the iso or initrd download handler.

## How was this code tested?


~~WIP: Still looking for a way to verify this manually but parsing out the concatenated initrd format seems non-trivial.~~
It was non-trivial but this is what I did:

- Configure this branch of the image service to run against the integration saas assisted service
- Create a cluster in integration UI and fetch the infra-env ID
- Download the pxe-initrd from the local image service
- Use [`binwalk`](https://www.kali.org/tools/binwalk/) on the resulting file to find the last gzipped archive location
```
$ binwalk 02bbb814-eafa-497b-811e-86931d697d33-initrd.img

DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
0             0x0             ASCII cpio archive (SVR4 with no CRC), file name: ".", file name length: "0x00000002", file size: "0x00000000"
112           0x70            ASCII cpio archive (SVR4 with no CRC), file name: "early_cpio", file name length: "0x0000000B", file size: "0x00000002"
240           0xF0            ASCII cpio archive (SVR4 with no CRC), file name: "kernel", file name length: "0x00000007", file size: "0x00000000"
360           0x168           ASCII cpio archive (SVR4 with no CRC), file name: "kernel/x86", file name length: "0x0000000B", file size: "0x00000000"
484           0x1E4           ASCII cpio archive (SVR4 with no CRC), file name: "kernel/x86/microcode", file name length: "0x00000015", file size: "0x00000000"
616           0x268           ASCII cpio archive (SVR4 with no CRC), file name: "kernel/x86/microcode/AuthenticAMD.bin", file name length: "0x00000026", file size: "0x00007752"
31312         0x7A50          ASCII cpio archive (SVR4 with no CRC), file name: "kernel/x86/microcode/GenuineIntel.bin", file name length: "0x00000026", file size: "0x005B9800"
6034148       0x5C12E4        ASCII cpio archive (SVR4 with no CRC), file name: "TRAILER!!!", file name length: "0x0000000B", file size: "0x00000000"
6034432       0x5C1400        gzip compressed data, maximum compression, from Unix, last modified: 1970-01-01 00:00:00 (null date)
14185156      0xD872C4        MySQL ISAM index file Version 3
20533552      0x1395130       gzip compressed data, from NTFS filesystem (NT), last modified: 1970-01-01 00:00:00 (null date)
20547306      0x13986EA       gzip compressed data, from NTFS filesystem (NT), last modified: 1970-01-01 00:00:00 (null date)
20551203      0x1399623       gzip compressed data, from NTFS filesystem (NT), last modified: 1970-01-01 00:00:00 (null date)
...
87837436      0x53C4AFC       gzip compressed data, has original file name: "dracut-random.cpio", from Unix, last modified: 2019-12-10 14:51:33
87837607      0x53C4BA7       gzip compressed data, maximum compression, from Unix, last modified: 1970-01-01 00:00:00 (null date)
87854752      0x53C8EA0       gzip compressed data, last modified: 1970-01-01 00:00:00 (null date)
```
- Extract the ignition and verify it is correct:
```
$ dd if=02bbb814-eafa-497b-811e-86931d697d33-initrd.img of=ignition.img.gz skip=87854752 bs=1
4144+0 records in
4144+0 records out
4144 bytes (4.1 kB, 4.0 KiB) copied, 0.0109975 s, 377 kB/s
$ file ignition.img.gz 
ignition.img.gz: gzip compressed data, original size modulo 2^32 10248
$ gunzip ignition.img.gz 
$ cpio -idmv --no-absolute-filenames < ignition.img 
config.ign
21 blocks
$ cat config.ign | jq
{
  "ignition": {
    "version": "3.1.0"
  },
  "passwd": {
    "users": []
  },
  "systemd": {
    "units": [
      {
        "name": "agent.service",
        "enabled": true,
...
```

## Assignees


/cc @CrystalChun 
/cc @filanov 
/cc @avishayt 

## Links


~~Depends on https://github.com/openshift/assisted-image-service/pull/62~~ Merged
Resolves https://issues.redhat.com/browse/MGMT-9333


## Checklist

- [x] Relevant issues have been associated
